zeppelin-zengine/src/main/java/org/apache/zeppelin/scheduler/RemoteScheduler.java

```diff
@@ -34,14 +36,14 @@
 public class RemoteScheduler extends AbstractScheduler {
   private static final Logger LOGGER = LoggerFactory.getLogger(RemoteScheduler.class);
 
-  private RemoteInterpreter remoteInterpreter;
-  private ExecutorService executor;
+  private final RemoteInterpreter remoteInterpreter;
+  private final ExecutorService executor;
 
   public RemoteScheduler(String name,
-                         ExecutorService executor,
                          RemoteInterpreter remoteInterpreter) {
     super(name);
-    this.executor = executor;
+    this.executor =
+        Executors.newSingleThreadExecutor(new SchedulerThreadFactory("FIFO-" + name + "-"));
```

Comment:
I try to replace this line with `Executors.newFixedThreadPool(10, new SchedulerThreadFactory(name));` and concurrent querys run normally. So there is two level scheduler: `RemoteScheduler` in zengine and another one in `JDBCInterpreter`. 

https://github.com/apache/zeppelin/blob/debfa7fcc23d6f5ea2f0664308cdbe7eba78b419/zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreter.java#L339-L355


https://github.com/apache/zeppelin/blob/47a86ac5fcd71f299e8bf605f25a070c478a2905/jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java#L1075-L1081

Reply:
IIRC, concurrent execution for jdbc interpreter is being handled by JDBC interpreter. As you mentioned, we also designed to select two types of interpreter for remote interpreter but it was not developed well in my understanding.

Reply:
Synchronization between the Zeppelin server and the remote interpreters is not easy. Currently, the Zeppelin server sends one paragraph per paragraph to a remote interpreter in order to delete the queue locally in the event of an abort.

A clean implementation would be needed so that the Zeppelin server can cancel jobs (paragraphs) in the remote interpreter. The Zeppelin server should only send several jobs (paragraphs) to the remote interpreter if such an implementation is available. The remote interpreter can then of course work in parallel or sequentially, depending on the actual interpreter and configuration.